Research/Research/AssessmentTaskObject.swift

```diff
@@ -0,0 +1,430 @@
+//
+// 1.  Redistributions of source code must retain the above copyright notice, this
+// list of conditions and the following disclaimer.
+//
+// 2.  Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation and/or
+// other materials provided with the distribution.
+//
+// 3.  Neither the name of the copyright holder(s) nor the names of any contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission. No license is granted to the trademarks of
+// the copyright holders even if such marks are included in this software.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+import Foundation
+import JsonModel
+
+
+/// The serialization "type" key for decoding of the task. This is used to decode the task using a
+/// `RSDFactory`.
+public struct RSDTaskType : TypeRepresentable, RawRepresentable, Codable, Hashable {
+    public let rawValue: String
+    
+    public init(rawValue: String) {
+        self.rawValue = rawValue
+    }
+
+    public static let assessment: RSDTaskType = "assessment"
+    
+    fileprivate static let null: RSDTaskType = "null"
+}
+
+extension RSDTaskType : ExpressibleByStringLiteral {
+    public init(stringLiteral value: String) {
+        self.init(rawValue: value)
+    }
+}
+
+extension RSDTaskType : DocumentableStringLiteral {
+    public static func examples() -> [String] {
+        [RSDTaskType.assessment.rawValue]
+    }
+}
+
+public final class TaskSerializer : AbstractPolymorphicSerializer, PolymorphicSerializer {
+    override init() {
+        let examples: [SerializableTask] = [
+            AssessmentTaskObject()
+        ]
+        self.examples = examples
+    }
+    
+    public private(set) var examples: [RSDTask]
+    
+    public func add(_ example: SerializableTask) {
+        if let idx = examples.firstIndex(where: {
+            ($0 as! PolymorphicRepresentable).typeName != example.typeName }) {
+            examples.remove(at: idx)
+        }
+        examples.append(example)
+    }
+}
+
+/// For `SerializableTask`, do not implement the typeName in an extension to allow the serializable
```

Comment:
Not really following the documentation here--maybe it makes more sense in context?

Reply:
Not really sure how else to explain it. There is nothing to keep a class from implementing both `RSDStep` and `RSDTask`. If you want to be able to deserialize as a `RSDStep` then you need to implement `var stepType: RSDStepType` and add to the step serializer. If you want to be able to deserialize as a `RSDTask` then you need to add it to the task serializer. You also need to implement `SerializableTask` and `SerializableStep`.  

`SerializableStep` includes this extension:

```
public extension SerializableStep {
    var typeName: String { return stepType.rawValue }
}
```

If I added an equivalent extension to `SerializableTask`:

```
public extension SerializableTask {
    var typeName: String { return taskType.rawValue }
}
```

*then* there would be a build conflict b/c Swift compiler wouldn't be able to determine *which* `typeName` method to call. Building would fail.

So I didn't do that.

Anyhow, I am not sure how to explain that in a code comment.